File: examples/vsliceJava/src/test/java/csw/examples/vsliceJava/assembly/ValidationTests.java
package csw.examples.vsliceJava.assembly;

import csw.services.ccs.CommandStatus;
import csw.util.config.Configurations;
import csw.util.config.Configurations.SequenceConfig;
import csw.util.config.Configurations.SetupConfig;
import csw.util.config.Configurations.SetupConfigArg;
import csw.util.config.DoubleKey;
import javacsw.services.ccs.JCommandStatus;
import org.junit.Test;

import java.util.List;
import java.util.stream.Collectors;

import static csw.examples.vsliceJava.assembly.AssemblyContext.*;
import static csw.examples.vsliceJava.assembly.ConfigValidation.*;
import static csw.services.ccs.Validation.*;
import static javacsw.services.ccs.JCommandStatus.NotAccepted;
import static javacsw.services.ccs.JValidation.Valid;
import static javacsw.services.ccs.JValidation.WrongConfigKeyIssue;
import static javacsw.util.config.JItems.jadd;
import static javacsw.util.config.JItems.jset;
import static javacsw.util.config.JUnitsOfMeasure.kilometers;
import static junit.framework.TestCase.*;


/**
 * These are tests of the calculations in the Calculation Actor
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public class ValidationTests {
  private static AssemblyContext assemblyContext = AssemblyTestData.TestAssemblyContext;

  Invalid checkInvalid(Validation result) {
    assertTrue(result instanceof Invalid);
    return (Invalid) result;
  }

  void checkForWrongConfigKey(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof WrongConfigKeyIssue);
  }

  void checkForMissingKeys(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof MissingKeyIssue);
  }

  void checkForWrongItemType(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof WrongItemTypeIssue);
  }

  void checkForWrongUnits(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof WrongUnitsIssue);
  }

  void checkForWrongNumberOfParameters(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof WrongNumberOfItemsIssue);
  }

  void checkForOutOfRange(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof ItemValueOutOfRangeIssue);
  }

  void checkForOtherIssue(Validation result) {
    assertTrue(checkInvalid(result).issue() instanceof OtherIssue);
  }

  /*
   * Test Description: This tests the validation of the init SC
   */

  // --- testing validation for init command ---

  @Test
  public void test1() {
    // should fail if not an init
    SetupConfig sc = new SetupConfig(assemblyContext.positionCK.prefix());
    checkForWrongConfigKey(initValidation(sc, assemblyContext));
  }

  @Test
  public void test2() {
    // should validate init setupconfig with 0 args
    SetupConfig sc = new SetupConfig(assemblyContext.initCK.prefix());

    // Should validate with no arguments
    assertEquals(initValidation(sc, assemblyContext), Valid);
  }

  @Test
  public void test3() {
    // should validate 2 arg init setupconfig
    SetupConfig sc = jadd(new SetupConfig(assemblyContext.initCK.prefix()),
      jset(configurationNameKey, "config1"),
      jset(configurationVersionKey, "1.0"));

    // Should validate with 2 good arguments
    assertEquals(initValidation(sc, assemblyContext), Valid);

    // Should be invalid with an extra argument
    sc = sc.add(jset(zenithAngleKey, 0.0));
    checkForWrongNumberOfParameters(initValidation(sc, assemblyContext));
  }

  @Test
  public void test4() {
    // should check for init item types
    // Make a key with the correct name that isn't the right type
    DoubleKey cvKey = new DoubleKey(configurationVersionKey.keyName());
    SetupConfig sc = jadd(new SetupConfig(assemblyContext.initCK.prefix()),
      jset(configurationNameKey, "config1"),
      jset(cvKey, 1.0));
    // Should be invalid
    checkForWrongItemType(initValidation(sc, assemblyContext));
  }

  /*
   * Test Description: This tests the validation of the move SC
   */

  // --- testing validation of move setupconfig ---

  // should fail if not a move
  @Test
  public void test5() {
    SetupConfig sc = new SetupConfig(assemblyContext.positionCK.prefix());
    checkForWrongConfigKey(moveValidation(sc, assemblyContext));
  }

  @Test
  public void test6() {
    // should validate the move setupconfig with 0 args
    SetupConfig sc = new SetupConfig(assemblyContext.moveCK.prefix());
    // Should validate with no arguments
    assertEquals(moveValidation(sc, assemblyContext), Valid);
  }

  @Test
  public void test7() {
    // should validate 1 arg move setupconfig
    // Create but don't set units
    SetupConfig sc = new SetupConfig(assemblyContext.moveCK.prefix()).add(jset(stagePositionKey, 22.0));

    // Should fail for units
    checkForWrongUnits(moveValidation(sc, assemblyContext));

    // Now add good units
    sc = sc.add(jset(stagePositionKey, 22.0).withUnits(stagePositionUnits));

    // Validates with 1 good argument
    assertEquals(moveValidation(sc, assemblyContext), Valid);

    // Should be valid with an extra argument in this case
    sc = sc.add(jset(zenithAngleKey, 0.0));
    assertEquals(moveValidation(sc, assemblyContext), Valid);
  }

  /*
   * Test Description: This tests the validation of the position SC
   */
  // --- testing validation of position setupconfig ---

  @Test
  public void test8() {
    // should fail if not a position
    SetupConfig sc = new SetupConfig(assemblyContext.moveCK.prefix());
    checkForWrongConfigKey(positionValidation(sc, assemblyContext));
  }

  @Test
  public void test9() {
    // should fail for missing unitsg
    SetupConfig sc = new SetupConfig(assemblyContext.positionCK.prefix()).add(jset(naRangeDistanceKey, 22.0));

    // Should fail for units
    checkForWrongUnits(positionValidation(sc, assemblyContext));
  }

  @Test
  public void test10() {
    // should validate when keys and units present
    // Now add good units
    SetupConfig sc = new SetupConfig(assemblyContext.positionCK.prefix()).add(
      jset(naRangeDistanceKey, 22.0).withUnits(naRangeDistanceUnits));

    // Should validate with 1 good argument
    assertEquals(positionValidation(sc, assemblyContext), Valid);

    // Should be valid with an extra argument in this case
    sc = sc.add(jset(zenithAngleKey, 0.0));
    assertEquals(positionValidation(sc, assemblyContext), Valid);
  }

  @Test
  public void test11() {
    // should fail for negative range distance value
    // Now  good units with neg value
    SetupConfig sc = new SetupConfig(assemblyContext.positionCK.prefix()).add(
      jset(naRangeDistanceKey, -22.0).withUnits(naRangeDistanceUnits));
    checkForOutOfRange(positionValidation(sc, assemblyContext));
  }

  /*
   * Test Description: This tests the validation of the setElevation SC
   */
  // --- testing validation for setElevation command ---

  @Test
  public void test12() {
    // should fail if not a setElevation
    SetupConfig sc = new SetupConfig(assemblyContext.initCK.prefix());
    checkForWrongConfigKey(setElevationValidation(sc, assemblyContext));
  }

  @Test
  public void test13() {
    // should vail to vailidate for missing units and keys
    // First check for missing args
    SetupConfig sc = new SetupConfig(assemblyContext.setElevationCK.prefix());
    checkForMissingKeys(setElevationValidation(sc, assemblyContext));

    // Should validate with 2 good arguments
    sc = jadd(sc, jset(zenithAngleKey, 0.0), jset(naElevationKey, 100.0));
    checkForWrongUnits(setElevationValidation(sc, assemblyContext));
  }

  @Test
  public void test14() {
    /// should validate 2 arg setElevation setupconfig
    SetupConfig sc = jadd(new SetupConfig(assemblyContext.setElevationCK.prefix()),
      jset(zenithAngleKey, 0.0).withUnits(zenithAngleUnits),
      jset(naElevationKey, 100.0).withUnits(naElevationUnits));
    assertEquals(setElevationValidation(sc, assemblyContext), Valid);

    // Should ignore an extra parameter
    sc = sc.add(jset(naRangeDistanceKey, 0.0));
    assertEquals(setElevationValidation(sc, assemblyContext), Valid);
  }

  @Test
  public void test15() {
    // should check for init item types
    // Make a key with the correct name that isn't the right type
    DoubleKey cvKey = new DoubleKey(configurationVersionKey.keyName());
    SetupConfig sc = jadd(new SetupConfig(assemblyContext.initCK.prefix()),
      jset(configurationNameKey, "config1"),
      jset(cvKey, 1.0));
    // Should be invalid
    assertTrue(initValidation(sc, assemblyContext) instanceof Invalid);
  }


  /*
   * Test Description: Test tests the validation of a setAngle SC
   */

  // --- testing validation of setAngle setupconfig

  public void test16() {
    // should fail if not a setAngle
    SetupConfig sc = new SetupConfig(assemblyContext.moveCK.prefix());
    checkForWrongConfigKey(setAngleValidation(sc, assemblyContext));
  }

  @Test
  public void test17() {
    // should fail for missing key
    SetupConfig sc = new SetupConfig(assemblyContext.setAngleCK.prefix());

    // Should fail for units
    checkForMissingKeys(setAngleValidation(sc, assemblyContext));
  }

  @Test
  public void test18() {
    // should fail for missing units
    SetupConfig sc = new SetupConfig(assemblyContext.setAngleCK.prefix()).add(jset(zenithAngleKey, 22.0));

    // Should fail for units
    checkForWrongUnits(setAngleValidation(sc, assemblyContext));
  }

  @Test
  public void test19() {
    // should validate when keys and units present
    // Now add good units
    SetupConfig sc = new SetupConfig(assemblyContext.setAngleCK.prefix()).add(
      jset(zenithAngleKey, 22.0).withUnits(zenithAngleUnits));

    // Should validate with 1 good argument
    assertEquals(setAngleValidation(sc, assemblyContext), Valid);

    // Should be valid with an extra argument in this case
    sc = sc.add(jset(naElevationKey, 0.0));
    assertEquals(setAngleValidation(sc, assemblyContext), Valid);
  }

  /*
   * Test Description: This tests the validation of a follow SC
   */
  // --- testing validation of follow setupconfig ---

  @Test
  public void test20() {
    // should fail if not a follow
    SetupConfig sc = new SetupConfig(assemblyContext.moveCK.prefix());
    checkForWrongConfigKey(followValidation(sc, assemblyContext));
  }

  @Test
  public void test21() {
    // should fail for missing key
    SetupConfig sc = new SetupConfig(assemblyContext.followCK.prefix());

    // Should fail for units
    checkForMissingKeys(followValidation(sc, assemblyContext));
  }

  @Test
  public void test22() {
    // should validate when key present
    SetupConfig sc = new SetupConfig(assemblyContext.followCK.prefix()).add(jset(nssInUseKey, true));

    // Should validate with 1 good argument
    assertEquals(followValidation(sc, assemblyContext), Valid);

    // Should be valid with an extra argument in this case
    sc = sc.add(jset(naElevationKey, 0.0));
    assertEquals(followValidation(sc, assemblyContext), Valid);
  }

  /*
   * Test Description: This is a test of the SetupConfigARg validation routine in TromboneAssembly
   */
  // --- Test of TromboneAssembly validation ---

  @Test
  public void test23() {
    // should work with okay sca
    SetupConfigArg sca = Configurations.createSetupConfigArg("testobsId",
      new SetupConfig(assemblyContext.initCK.prefix()),
      new SetupConfig(assemblyContext.stopCK.prefix()));

    List<Invalid> issues = invalidsInTromboneSetupConfigArg(sca, assemblyContext);
    assertTrue(issues.isEmpty());
  }

  @Test
  public void test24() {
    // should show a single issue
    // positionCK requires an argument
    SetupConfigArg sca = Configurations.createSetupConfigArg("testobsId",
      new SetupConfig(assemblyContext.initCK.prefix()),
      new SetupConfig(assemblyContext.positionCK.prefix()));
    List<Invalid> issues = invalidsInTromboneSetupConfigArg(sca, assemblyContext);
    assertFalse(issues.isEmpty());
    assertEquals(issues.size(), 1);
    checkForMissingKeys(issues.get(0));
  }

  @Test
  public void test25() {
    // should show multiple issues
    // positionCK needs an argument and moveCK has the wrong units
    SetupConfigArg sca = Configurations.createSetupConfigArg("testobsId",
      new SetupConfig(assemblyContext.initCK.prefix()),
      new SetupConfig(assemblyContext.positionCK.prefix()),
      jadd(new SetupConfig(assemblyContext.moveCK.prefix()), jset(stagePositionKey, 22.0).withUnits(kilometers)));
    List<Invalid> issues = invalidsInTromboneSetupConfigArg(sca, assemblyContext);
    assertTrue(!issues.isEmpty());
    assertEquals(issues.size(), 2);
    checkForMissingKeys(issues.get(0));
    checkForWrongUnits(issues.get(1));
  }

  @Test
  public void test26() {
    // should convert validation invalid successfully to a CommandStatus invalid
    String testmessage = "test message";

    Invalid t1 = new Invalid(WrongConfigKeyIssue(testmessage));

    CommandStatus.Invalid c1 = JCommandStatus.Invalid(t1);
    assertTrue(c1.issue() instanceof WrongConfigKeyIssue);
    assertEquals(c1.issue().reason(), testmessage);
  }

  @Test
  public void test27() {
    // should convert validation result to comand status result
    SetupConfigArg sca = Configurations.createSetupConfigArg("testobsId",
      new SetupConfig(assemblyContext.initCK.prefix()),
      new SetupConfig(assemblyContext.positionCK.prefix()),
      new SetupConfig(assemblyContext.moveCK.prefix()).add(jset(stagePositionKey, 22.0).withUnits(kilometers)));

    // Check if validated properly
    List<Validation> validations = ConfigValidation.validateTromboneSetupConfigArg(sca, assemblyContext);
    assertEquals(validations.size(), sca.configs().size());
    assertEquals(validations.get(0), Valid);
    assertTrue(validations.get(1) instanceof Invalid);
    assertTrue(validations.get(2) instanceof Invalid);

    // Convert to pairs
    List<SequenceConfig> configs = sca.getConfigs().stream().map(f -> (SequenceConfig) f).collect(Collectors.toList());
    List<CommandStatus.CommandResultPair> cresult = CommandStatus.validationsToCommandResultPairs(configs, validations);
    assertEquals(cresult.size(), sca.configs().size());
    assertEquals(cresult.get(0).status(), JCommandStatus.Valid);
    assertEquals(cresult.get(0).config(), sca.getConfigs().get(0));

    assertTrue(cresult.get(1).status() instanceof CommandStatus.Invalid);
    assertEquals(cresult.get(1).config(), sca.getConfigs().get(1));

    assertTrue(cresult.get(2).status() instanceof CommandStatus.Invalid);
    assertEquals(cresult.get(2).config(), sca.getConfigs().get(2));

    // Is correct overall returned
    assertEquals(CommandStatus.validationsToOverallCommandStatus(validations), NotAccepted);

    // Same with no errors
    SetupConfigArg sca2 = Configurations.createSetupConfigArg("testobsId",
      new SetupConfig(assemblyContext.initCK.prefix()), assemblyContext.positionSC(22.0), assemblyContext.moveSC(44.0));

    List<Validation> validations2 = ConfigValidation.validateTromboneSetupConfigArg(sca2, assemblyContext);
    assertTrue(isAllValid(validations2));
  }
}

File: examples/vslice/src/test/scala/csw/examples/vslice/assembly/ValidationTests.scala
package csw.examples.vslice.assembly

import csw.services.ccs.CommandStatus
import csw.services.ccs.CommandStatus.NotAccepted
import csw.services.ccs.Validation._
import csw.util.config.Configurations.SetupConfig
import csw.util.config.UnitsOfMeasure.kilometers
import csw.util.config.{Configurations, DoubleKey}
import org.scalatest.{BeforeAndAfterAll, FunSpec, Inspectors, ShouldMatchers}

/**
 * TMT Source Code: 8/25/16.
 */
class ValidationTests extends FunSpec with ShouldMatchers with Inspectors with BeforeAndAfterAll {
  import ConfigValidation._

  implicit val ac = AssemblyTestData.TestAssemblyContext
  import ac._

  def checkInvalid(result: Validation): Invalid = {
    result shouldBe a[Invalid]
    result.asInstanceOf[Invalid]
  }

  def checkForWrongConfigKey(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[WrongConfigKeyIssue]
  }

  def checkForMissingKeys(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[MissingKeyIssue]
  }

  def checkForWrongItemType(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[WrongItemTypeIssue]
  }

  def checkForWrongUnits(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[WrongUnitsIssue]
  }

  def checkForWrongNumberOfParameters(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[WrongNumberOfItemsIssue]
  }

  def checkForOutOfRange(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[ItemValueOutOfRangeIssue]
  }

  def checkForOtherIssue(result: Validation): Unit = {
    checkInvalid(result).issue shouldBe a[OtherIssue]
  }

  /**
   * Test Description: This tests the validation of the init SC
   */
  describe("testing validation for init command") {

    it("should fail if not an init") {
      val sc = SetupConfig(positionCK)
      checkForWrongConfigKey(initValidation(sc))
    }

    it("should validate init setupconfig with 0 args") {
      val sc = SetupConfig(initCK)

      // Should validate with no arguments
      initValidation(sc) should be(Valid)
    }
    it("should validate 2 arg init setupconfig") {
      var sc = SetupConfig(initCK).madd(configurationNameKey -> "config1", configurationVersionKey -> "1.0")

      // Should validate with 2 good arguments
      initValidation(sc) should be(Valid)

      // Should be invalid with an extra argument
      sc = sc.add(zenithAngleKey -> 0.0)
      checkForWrongNumberOfParameters(initValidation(sc))
      //initValidation(sc).isInstanceOf[Invalid] should be(true)
    }
    it("should check for init item types") {
      // Make a key with the correct name that isn't the right type
      val cvKey = DoubleKey(configurationVersionKey.keyName)
      val sc = SetupConfig(initCK).madd(configurationNameKey -> "config1", cvKey -> 1.0)
      // Should be invalid
      checkForWrongItemType(initValidation(sc))
    }
  }

  /**
   * Test Description: This tests the validation of the move SC
   */
  describe("testing validation of move setupconfig") {
    it("should fail if not a move") {
      val sc = SetupConfig(positionCK)
      checkForWrongConfigKey(moveValidation(sc))
    }

    it("should validate the move setupconfig with 0 args") {
      val sc = SetupConfig(moveCK)
      // Should validate with no arguments
      moveValidation(sc) should be(Valid)
    }

    it("should validate 1 arg move setupconfig") {
      // Create but don't set units
      var sc = SetupConfig(moveCK).add(stagePositionKey -> 22.0)

      // Should fail for units
      checkForWrongUnits(moveValidation(sc))

      // Now add good units
      sc = sc.add(stagePositionKey -> 22.0 withUnits stagePositionUnits)

      // Validates with 1 good argument
      moveValidation(sc) should be(Valid)

      // Should be valid with an extra argument in this case
      sc = sc.add(zenithAngleKey -> 0.0)
      moveValidation(sc) shouldBe Valid
    }
  }

  /**
   * Test Description: This tests the validation of the position SC
   */
  describe("testing validation of position setupconfig") {
    it("should fail if not a position") {
      val sc = SetupConfig(moveCK)
      checkForWrongConfigKey(positionValidation(sc))
    }

    it("should fail for missing unitsg") {
      val sc = SetupConfig(positionCK).add(naRangeDistanceKey -> 22.0)

      // Should fail for units
      checkForWrongUnits(positionValidation(sc))
    }

    it("should validate when keys and units present") {
      // Now add good units
      var sc = SetupConfig(positionCK).add(naRangeDistanceKey -> 22.0 withUnits naRangeDistanceUnits)

      // Should validate with 1 good argument
      positionValidation(sc) shouldBe Valid

      // Should be valid with an extra argument in this case
      sc = sc.add(zenithAngleKey -> 0.0)
      positionValidation(sc) shouldBe Valid
    }

    it("should fail for negative range distance value") {
      // Now  good units with neg value
      val sc = SetupConfig(positionCK).add(naRangeDistanceKey -> -22.0 withUnits naRangeDistanceUnits)
      checkForOutOfRange(positionValidation(sc))
    }
  }

  /**
   * Test Description: This tests the validation of the setElevation SC
   */
  describe("testing validation for setElevation command") {

    it("should fail if not a setElevation") {
      val sc = SetupConfig(initCK)
      checkForWrongConfigKey(setElevationValidation(sc))
    }

    it("should vail to vailidate for missing units and keys") {
      // First check for missing args
      var sc = SetupConfig(setElevationCK)
      checkForMissingKeys(setElevationValidation(sc))

      // Should validate with 2 good arguments
      sc = sc.madd(zenithAngleKey -> 0.0, naElevationKey -> 100.0)
      checkForWrongUnits(setElevationValidation(sc))
    }

    it("should validate 2 arg setElevation setupconfig") {
      var sc = SetupConfig(setElevationCK).madd(zenithAngleKey -> 0.0 withUnits zenithAngleUnits, naElevationKey -> 100.0 withUnits naElevationUnits)
      setElevationValidation(sc) should be(Valid)

      // Should ignore an extra parameter
      sc = sc.add(naRangeDistanceKey -> 0.0)
      setElevationValidation(sc) shouldBe Valid
    }

    it("should check for init item types") {
      // Make a key with the correct name that isn't the right type
      val cvKey = DoubleKey(configurationVersionKey.keyName)
      val sc = SetupConfig(initCK).madd(configurationNameKey -> "config1", cvKey -> 1.0)
      // Should be invalid
      //val result = initValidation(sc)
      //info("result: " + result)
      initValidation(sc).isInstanceOf[Invalid] should be(true)
    }
  }

  /**
   * Test Description: Test tests the validation of a setAngle SC
   */
  describe("testing validation of setAngle setupconfig") {
    it("should fail if not a setAngle") {
      val sc = SetupConfig(moveCK)
      checkForWrongConfigKey(setAngleValidation(sc))
    }

    it("should fail for missing key") {
      val sc = SetupConfig(setAngleCK)

      // Should fail for units
      checkForMissingKeys(setAngleValidation(sc))
    }

    it("should fail for missing units") {
      val sc = SetupConfig(setAngleCK).add(zenithAngleKey -> 22.0)

      // Should fail for units
      checkForWrongUnits(setAngleValidation(sc))
    }

    it("should validate when keys and units present") {
      // Now add good units
      var sc = SetupConfig(setAngleCK).add(zenithAngleKey -> 22.0 withUnits zenithAngleUnits)

      // Should validate with 1 good argument
      setAngleValidation(sc) should be(Valid)

      // Should be valid with an extra argument in this case
      sc = sc.add(naElevationKey -> 0.0)
      setAngleValidation(sc) shouldBe Valid
    }
  }

  /**
   * Test Description: This tests the validation of a follow SC
   */
  describe("testing validation of follow setupconfig") {
    it("should fail if not a follow") {
      val sc = SetupConfig(moveCK)
      checkForWrongConfigKey(followValidation(sc))
    }

    it("should fail for missing key") {
      val sc = SetupConfig(followCK)

      // Should fail for units
      checkForMissingKeys(followValidation(sc))
    }

    it("should validate when key present") {
      var sc = SetupConfig(followCK).add(nssInUseKey -> true)

      // Should validate with 1 good argument
      followValidation(sc) should be(Valid)

      // Should be valid with an extra argument in this case
      sc = sc.add(naElevationKey -> 0.0)
      followValidation(sc) shouldBe Valid
    }
  }

  /**
   * Test Description: This is a test of the SetupConfigARg validation routine in TromboneAssembly
   */
  describe("Test of TromboneAssembly validation") {
    //implicit val tc = AssemblyTestData.TestAssemblyContext

    it("should work with okay sca") {
      val sca = Configurations.createSetupConfigArg("testobsId", SetupConfig(initCK), SetupConfig(stopCK))

      val issues = invalidsInTromboneSetupConfigArg(sca)
      issues shouldBe empty
    }

    it("should show a single issue") {
      // positionCK requires an argument
      val sca = Configurations.createSetupConfigArg("testobsId", SetupConfig(initCK), SetupConfig(positionCK))
      val issues: Seq[Invalid] = invalidsInTromboneSetupConfigArg(sca)
      issues should not be empty
      issues.size should be(1)
      checkForMissingKeys(issues.head)
    }

    it("should show multiple issues") {
      // positionCK needs an argument and moveCK has the wrong units
      val sca = Configurations.createSetupConfigArg(
        "testobsId",
        SetupConfig(initCK),
        SetupConfig(positionCK),
        SetupConfig(moveCK).add(stagePositionKey -> 22 withUnits kilometers)
      )
      val issues = invalidsInTromboneSetupConfigArg(sca)
      issues should not be empty
      issues.size should be(2)
      checkForMissingKeys(issues.head)
      checkForWrongUnits(issues(1))
    }

    it("should convert validation invalid successfully to a CommandStatus invalid") {
      //import csw.services.ccs.CommandStatus.Invalid
      val testmessage = "test message"

      val t1 = Invalid(WrongConfigKeyIssue(testmessage))

      val c1 = CommandStatus.Invalid(t1)
      c1.issue shouldBe a[WrongConfigKeyIssue]
      c1.issue.reason should equal(testmessage)

    }

    it("should convert validation result to comand status result") {
      val sca = Configurations.createSetupConfigArg("testobsId", SetupConfig(initCK), SetupConfig(positionCK), SetupConfig(moveCK).add(stagePositionKey -> 22 withUnits kilometers))

      // Check if validated properly
      val validations = ConfigValidation.validateTromboneSetupConfigArg(sca)
      validations.size should equal(sca.configs.size)
      validations.head shouldBe Valid
      validations(1) shouldBe a[Invalid]
      validations(2) shouldBe a[Invalid]

      // Convert to pairs
      val cresult = CommandStatus.validationsToCommandResultPairs(sca.configs, validations)
      cresult.size should equal(sca.configs.size)
      cresult.head.status shouldBe CommandStatus.Valid
      cresult.head.config should equal(sca.configs.head)

      cresult(1).status shouldBe a[CommandStatus.Invalid]
      cresult(1).config should equal(sca.configs(1))

      cresult(2).status shouldBe a[CommandStatus.Invalid]
      cresult(2).config should equal(sca.configs(2))

      // Is correct overall returned
      CommandStatus.validationsToOverallCommandStatus(validations) shouldBe NotAccepted

      // Same with no errors
      val sca2 = Configurations.createSetupConfigArg("testobsId", SetupConfig(initCK), positionSC(22.0), moveSC(44.0))

      val validations2 = ConfigValidation.validateTromboneSetupConfigArg(sca2)
      isAllValid(validations2) shouldBe true

    }

  }

}

